### ISSUE TITLE
MLJ traits for GMMClusterer

### ISSUE THREAD

**Issue by juliohm** [sylvaticus/BetaML.jl#26]
The GMMClusterer is an unsupervised probabilistic model. However we can't check that programmatically because of https://github.com/JuliaAI/MLJModelInterface.jl/issues/120

Is there any fix to make sure that both KMeans and GMMClusterer return a set of categorical values? Right now `predict(Kmeans(), ...)` will return a vector of categorical values whereas `predict(GMMClusterer(), ...)` will return a vector of distributions.

**Comment by sylvaticus** [comment]
Hi, I am open to change the structure of the output of the MLJ interface of BetaML, however I would like to have the opinion of the MLJ maintainers... (@ablaom  ...)

The main difference between the two is that KMeans is a hard-culstering algorithm, while GMM is a soft-clustering, that is it reports the probability of the points to be in the different classes. I believe it is important to keep this information in the output...

**Comment by juliohm** [comment]
I think my proposal is to encode this difference explicitly in the trait system instead of in the output. The current trait system is not exhaustive in the sense that there isn't a Unsupervised + Probabilistic model in the type hierarchy.

**Comment by ablaom** [comment]
@sylvaticus There is a trait called `prediction_type` which should to `:probabilistic`. You can simply add the definition [here](https://github.com/sylvaticus/BetaML.jl/blob/8de3347bc4d8519ea7d2bcbbaf9108dcec5bdd33/src/Clustering/Clustering_MLJ.jl#L221). I don't see a need for any other change to your implementation. 

> Is there any fix to make sure that both KMeans and GMMClusterer return a set of categorical values?

@juliohm Apart from just broadcasting `mode` on the result, there are a couple of ways should work, but don't. I've opened [this issue](https://github.com/JuliaAI/MLJBase.jl/issues/656). Thanks for raising this point.

**Comment by sylvaticus** [comment]
Sorry for the delay, I have been very busy.
I have tried to implement the `prediction_type  = :probabilistic` on `MMI.metadata_model(GMMClusterer,...)` but I receive the error that it doesn't know about that keyword:

```
ERROR: LoadError: LoadError: LoadError: MethodError: no method matching metadata_model(::Type{BetaML.Clustering.GMMClusterer}; input_scitype=ScientificTypes.Table{var"#s46"} where var"#s46"<:(AbstractVector{var"#s9"} where var"#s9"<:Union{Missing, ScientificTypes.Continuous}), output_scitype=AbstractArray{var"#s398", N} where {var"#s398"<:ScientificTypes.Multiclass, N}, target_scitype=AbstractArray{var"#s397", N} where {var"#s397"<:ScientificTypes.Multiclass, N}, prediction_type=:probabilistic, supports_weights=false, descr="A Expectation-Maximisation clustering algorithm with customisable mixtures, from the Beta Machine Learning Toolkit (BetaML).", load_path="BetaML.Clustering.GMMClusterer")
Closest candidates are:
  metadata_model(::Any; input, target, output, weights, descr, path, input_scitype, target_scitype, output_scitype, supports_weights, docstring, load_path) at /home/lobianco/.julia/packages/MLJModelInterface/7iUwG/src/metadata_utils.jl:101 got unsupported keyword argument "prediction_type"
```

@ablaom When `:prediciton_type` has been added to MLJ ?  Is it a problem I am using old versions ?

**Comment by ablaom** [comment]
I'm guessing you are using the convenience function `metadata_model` to set this trait? It has not been added there, so you need define explicitly. Something like

```julia
MLJModelInterface.prediction_type(::Type{<:GMMClusterer}) = :probabilistic
```

Does that help?

**Comment by sylvaticus** [comment]
Thanks, implemented in commit bd489900384e3808fa5d8c7271ae377cdec3fa5a
